torch/distributed/device_mesh.py

```diff
@@ -392,81 +95,123 @@ def _get_device_handle(device_type: str = "cuda"):
 
+                sub_size = sub_size // new_size
+                sub_stride = ceil_div(sub_stride, curr_size)
+
+            if sub_size != 1 or len(sub_res.sizes_and_strides) == 0:
+                sub_res.sizes_and_strides.append((sub_size, sub_stride * l.strides[-1]))
+
+            res.append(sub_res)
+
+        return res
+
+    # Borrowed from pycute.layout.complement
+    # https://github.com/NVIDIA/cutlass/blob/6dd13d42784ee5bfa232d2441e6b9a021c5c6290/python/pycute/layout.py#L232
+    def complement(l: Layout) -> Layout:
+        res = Layout([])
+        current_idx = 1
+
+        for size, stride in sorted(l.sizes_and_strides, key=lambda x: x[1]):
+            if stride == 0 or size == 1:
+                continue
+
+            assert current_idx <= size * stride
+
+            res.sizes_and_strides.append((stride // current_idx, current_idx))
+            current_idx = size * stride
+
+        res.sizes_and_strides.append(
+            (ceil_div(get_world_size(), current_idx), current_idx)
+        )
+
+        return coalesce(res)
+
+    def rank_tensor_to_layouts(mesh: torch.Tensor) -> list[Layout]:
+        return [Layout([(mesh.size(i), mesh.stride(i))]) for i in range(mesh.ndim)]
+
+    def layout_to_rank_group_0(l: Layout) -> list[int]:
+        return [
+            sum(c * s for c, s in zip(coord, l.strides))
+            for coord in product(*(range(s) for s in l.sizes))
+        ]
+
+    def layout_to_all_rank_groups(l: Layout) -> list[list[int]]:
+        return [
+            [outer_offset + inner_offset for inner_offset in layout_to_rank_group_0(l)]
+            for outer_offset in layout_to_rank_group_0(complement(l))
+        ]
+
+    class DeviceMeshStorage:
         def __init__(
             self,
             device_type: str,
-            mesh: Union[torch.Tensor, "ArrayLike"],
-            *,
-            mesh_dim_names: Optional[tuple[str, ...]] = None,
-            backend_override: Optional[
-                tuple[tuple[Optional[str], Optional[C10dBackend.Options]], ...]
-            ] = None,
             _init_backend: bool = True,
         ) -> None:
             self.device_type = device_type
-            if isinstance(mesh, torch.Tensor) and mesh.device.type != "cpu":
-                raise ValueError(f"`mesh` must be a CPU tensor, got {mesh}")
-            self.mesh = (
-                mesh.detach().to(dtype=torch.int)
-                if isinstance(mesh, torch.Tensor)
-                else torch.tensor(mesh, device="cpu", dtype=torch.int)
-            )
-            self.mesh_dim_names = tuple(mesh_dim_names) if mesh_dim_names else None
-            if backend_override is None:
-                backend_override = ((None, None),) * self.mesh.ndim
-
-            # private field to pre-generate DeviceMesh's hash
-            self._flatten_mesh_list = tuple(self.mesh.flatten().tolist())
-            self._thread_id = None
+            self.layouts_to_groups: dict[Layout, ProcessGroup] = {}
```

Comment:
Sorry I forgot to mention you probably cannot directly store ProcessGroup here because it not serialiserble and it causes problem for checkpointing, that's why originally we use pg_name as a unique identifier for each PG.. this is kind of unavoidable? IDK, we are open for better suggestions.

Reply:
Uhm, in which cases do people attempt to serialize ProcessGroups? I can see how this might accidentally be serialized "transitively" if someone dumps a DTensor, which pulls in its DeviceMesh, which would have a reference to this cache. But regardless of whether we serialize the group objects or just their names, reloading this data would _not_ recreate the groups, right?

I suppose that if we end up having a single global singleton cache object then this problem might go away because each DeviceMesh would _not_ store an explicit reference to the cache, hence the transitive chain would be broken. Right?

Reply:
DCP doesn't serialize DTensor but some people who don't want to do resharding or use any other fancy techniques do use `torch.save(DTensor)`. This has been happening internally and externally. I'm not sure if people eventually change to DCP, but this serves an easy approach for users to explore DTensor adoption without learning another library, DCP. cc., @wconstab